### PR TITLE
sgf-parsing: Add individual tests of escaping/whitespace behaviour

### PR DESCRIPTION
**Pull request by petertseng**
11c36323-93fc-495d-bb23-c88ee5844b8c crams too much into one test case.
Split it up into multiple ones.

In addition, 11c36323-93fc-495d-bb23-c88ee5844b8c has behaviour that
violates the specification. The violation is that `\t` and `\n` (written
as `"\\t"` and `"\\n"` in the JSON string, respectively), do not hold
any sort of special significance in SGF, according to the specification:
https://www.red-bean.com/sgf/sgf4.html

Reimplement in 08e4b8ba-bb07-4431-a3d9-b1f4cdea6dab.

The reimplemented case is mostly as it was when the exercise was
originally implemented in
https://github.com/exercism/exercism/commit/7a5075bf96e2cafb29f2f883ff98b708be5594a6
and the reimplemented case is in accordance with the specification.

Note that the original case also got it wrong in that newlines should
remain newlines; this is corrected in
08e4b8ba-bb07-4431-a3d9-b1f4cdea6dab.

### CODE REVIEW

**Comment by petertseng** [comment]
Diff to https://github.com/petertseng/exercism-problem-specifications/blob/verify/exercises/sgf-parsing/verify.rb:

```diff
diff --git a/exercises/sgf-parsing/verify.rb b/exercises/sgf-parsing/verify.rb
index 45d41aca..f85c5044 100644
--- a/exercises/sgf-parsing/verify.rb
+++ b/exercises/sgf-parsing/verify.rb
@@ -97,12 +97,8 @@ class SGF
       if c == ?\\ && !escape
         escape = true
       else
-        val << ?\\ if escape && c == ?n
-        if escape && c == ?t
-          val << ' '
-        else
-          val << c
-        end
+        to_insert = c == "\n" && escape ? '' : c == "\t" ? ' ' : c
+        val << to_insert
         escape = false
       end
       pos += 1
@@ -114,6 +110,8 @@ end
 
 json = JSON.parse(File.read(File.join(__dir__, 'canonical-data.json')))
 
+json['cases'].reject! { |v| v['uuid'] == '11c36323-93fc-495d-bb23-c88ee5844b8c' }
+
 verify(json['cases'], property: 'parse') { |i, c|
   pos, node = SGF.new(i['encoded']).node(0, paren_required: true)
   raise 'Incomplete parse' unless pos == i['encoded'].size

```

File: exercises/sgf-parsing/canonical-data.json [hunk 1]
```diff
@@ -170,6 +170,171 @@
...
+            "uuid": "dc13ca67-fac0-4b65-b3fe-c584d6a2c523",
+            "description": "Opening bracket in property value doesn't need to be escaped",
+            "property": "parse",
+            "input": {
+                "encoded": "(;A[x[y\\]z]B[foo];C[bar])"
+            },
+            "expected": {
+                "properties": {
+                    "A": ["x[y]z"],
+                    "B": ["foo"]
```
Comment (petertseng) [inline exercises/sgf-parsing/canonical-data.json:239]: you may ask. why the `foo` and `bar`? Just to detect the case where a parser silently breaks in a way that finishes the current property value but halts all further parsing. Having additional material to parse after that, and verifying that that was parsed, avoids that case.

File: exercises/sgf-parsing/canonical-data.json [hunk 2]
```diff
@@ -170,6 +170,171 @@
...
+            "uuid": "08e4b8ba-bb07-4431-a3d9-b1f4cdea6dab",
+            "reimplements": "11c36323-93fc-495d-bb23-c88ee5844b8c",
+            "description": "mixing various kinds of whitespace and escaped characters in property value",
+            "property": "parse",
+            "input": {
+                "encoded": "(;A[\\]b\nc\\\nd\t\te\\\\ \\\n\\]])"
+            },
+            "expected": {
+                "properties": {
+                    "A": ["]b\ncd  e\\ ]"]
```
Comment (petertseng) [inline exercises/sgf-parsing/canonical-data.json:333]: To be very clear: In this case, and in all other cases in this PR as well, it's important to keep in mind the difference between the string as it appears in JSON and the string as it is presented to the SGF parser.

The string `"(;A[\\]b\nc\\\nd\t\te\\\\ \\\n\\]])"` in JSON means that this string is presented to the SGF parser:

```text
(;A[\]b
c\
d		e\\ \
\]])
```

Everything between the `[` and the `]` is the property value, and it contains these characters:

* `\]`, escaped closing bracket: insert closing bracket in property value
* `b`: insert as-is.
* unescaped newline: insert as-is.
* `c`: insert as-is.
* escaped newline: insert nothing.
* `d`: insert as-is
* two tabs: insert two spaces
* `e`: insert as-is
* `\\`, escaped backslash: insert backslash
* space: insert as-is.
* escaped newline: insert nothing
* `\]`, escaped closing bracket: insert closing bracket

The property value is thus:

```text
]b
cd  e\ ]
```

which is written in JSON as it is on line 333.

File: exercises/sgf-parsing/canonical-data.json [hunk 3]
```diff
@@ -170,6 +170,185 @@
...
+            "expected": {
+                "properties": {
+                    "A": ["hello\n\nworld"]
+                },
+                "children": []
+            }
+        },
+        {
+            "uuid": "8e4c970e-42d7-440e-bfef-5d7a296868ef",
+            "description": "Escaped closing bracket within property value becomes just a closing bracket",
```
Comment (ErikSchierboom) [inline exercises/sgf-parsing/canonical-data.json:203]: For consistency with the other test cases:

```suggestion
            "description": "escaped closing bracket within property value becomes just a closing bracket",
```

File: exercises/sgf-parsing/canonical-data.json [hunk 4]
```diff
@@ -170,6 +170,185 @@
...
+            "expected": {
+                "properties": {
+                    "A": ["]"]
+                },
+                "children": []
+            }
+        },
+        {
+            "uuid": "cf371fa8-ba4a-45ec-82fb-38668edcb15f",
+            "description": "Escaped backslash in property value becomes just a backslash",
```
Comment (ErikSchierboom) [inline exercises/sgf-parsing/canonical-data.json:217]: For consistency with the other test cases:

```suggestion
            "description": "escaped backslash in property value becomes just a backslash",
```

File: exercises/sgf-parsing/canonical-data.json [hunk 5]
```diff
@@ -170,6 +170,185 @@
...
+            "expected": {
+                "properties": {
+                    "A": ["\\"]
+                },
+                "children": []
+            }
+        },
+        {
+            "uuid": "dc13ca67-fac0-4b65-b3fe-c584d6a2c523",
+            "description": "Opening bracket within property value doesn't need to be escaped",
```
Comment (ErikSchierboom) [inline exercises/sgf-parsing/canonical-data.json:231]: For consistency with the other test cases:

```suggestion
            "description": "opening bracket within property value doesn't need to be escaped",
```

File: exercises/sgf-parsing/canonical-data.json [hunk 6]
```diff
@@ -170,6 +170,185 @@
...
+                            "C": ["baz"]
+                        },
+                        "children": []
+                    }
+                ]
+            }
+        },
+        {
+            "uuid": "a780b97e-8dbb-474e-8f7e-4031902190e8",
+            "description": "Semicolon in property value doesn't need to be escaped",
```
Comment (ErikSchierboom) [inline exercises/sgf-parsing/canonical-data.json:253]: For consistency with the other test cases:

```suggestion
            "description": "semicolon in property value doesn't need to be escaped",
```

File: exercises/sgf-parsing/canonical-data.json [hunk 7]
```diff
@@ -170,6 +170,185 @@
...
+                            "C": ["baz"]
+                        },
+                        "children": []
+                    }
+                ]
+            }
+        },
+        {
+            "uuid": "0b57a79e-8d89-49e5-82b6-2eaaa6b88ed7",
+            "description": "Parentheses in property value don't need to be escaped",
```
Comment (ErikSchierboom) [inline exercises/sgf-parsing/canonical-data.json:275]: For consistency with the other test cases:

```suggestion
            "description": "parentheses in property value don't need to be escaped",
```

File: exercises/sgf-parsing/canonical-data.json [hunk 8]
```diff
@@ -170,6 +170,185 @@
...
+                            "C": ["baz"]
+                        },
+                        "children": []
+                    }
+                ]
+            }
+        },
+        {
+            "uuid": "c72a33af-9e04-4cc5-9890-1b92262813ac",
+            "description": "Escaped tab in property value is converted to space",
```
Comment (ErikSchierboom) [inline exercises/sgf-parsing/canonical-data.json:297]: For consistency with the other test cases:

```suggestion
            "description": "escaped tab in property value is converted to space",
```

File: exercises/sgf-parsing/canonical-data.json [hunk 9]
```diff
@@ -170,6 +170,185 @@
...
+            "expected": {
+                "properties": {
+                    "A": ["hello world"]
+                },
+                "children": []
+            }
+        },
+        {
+            "uuid": "3a1023d2-7484-4498-8d73-3666bb386e81",
+            "description": "Escaped newline in property value is converted to nothing at all",
```
Comment (ErikSchierboom) [inline exercises/sgf-parsing/canonical-data.json:311]: For consistency with the other test cases:

```suggestion
            "description": "escaped newline in property value is converted to nothing at all",
```

File: exercises/sgf-parsing/canonical-data.json [hunk 10]
```diff
@@ -170,6 +170,185 @@
...
+            "expected": {
+                "properties": {
+                    "A": ["helloworld"]
+                },
+                "children": []
+            }
+        },
+        {
+            "uuid": "25abf1a4-5205-46f1-8c72-53273b94d009",
+            "description": "Escaped t and n in property value are just letters, not whitespace",
```
Comment (ErikSchierboom) [inline exercises/sgf-parsing/canonical-data.json:325]: For consistency with the other test cases:

```suggestion
            "description": "escaped t and n in property value are just letters, not whitespace",
```

File: exercises/sgf-parsing/canonical-data.json (hunk 2 above)
Comment (ErikSchierboom) [reply exercises/sgf-parsing/canonical-data.json:347]: Would it perhaps be worth it to add this comment to the test case? It looks to be very useful in getting people to understand what they're working with.

File: exercises/sgf-parsing/canonical-data.json (hunk 9 above)
Comment (ErikSchierboom) [reply exercises/sgf-parsing/canonical-data.json:311]: Would it perhaps make sense to have this test case moved up above the previous test case, as that test case also depends on escaped newlines being converted to nothing at all?

Comment (petertseng) [reply exercises/sgf-parsing/canonical-data.json]: For this particular case (3a1023d2-7484-4498-8d73-3666bb386e81 "escaped newline in property value is converted to nothing at all", this is the first case I see in this file that has an escaped newline) - am I thinking of a different case than the one you were talking about?

File: exercises/sgf-parsing/canonical-data.json (hunk 2 above)
Comment (petertseng) [reply exercises/sgf-parsing/canonical-data.json:365]: Not a bad idea. Added.

File: exercises/sgf-parsing/canonical-data.json [hunk 11]
```diff
@@ -170,6 +170,203 @@
...
+                "children": []
+            }
+        },
+        {
+            "comments": [
+                "Keep in mind the difference between:",
+                "1. the string as it is represented in a string literal",
+                "2. the string as it will be presented to the SGF parser",
+                "In particular, the SGF parser will see a property (between the square brackets) with these characters:",
+                "Escaped closing bracket (a single backslash followed by a closing bracket): Insert closing bracket in property value",
```
Comment (petertseng) [inline exercises/sgf-parsing/canonical-data.json:343]: this is verbose, but I am thinking it may be a good idea to avoid backslashes in the comments field if it will prevent further confusion.

File: exercises/sgf-parsing/canonical-data.json (hunk 9 above)
Comment (ErikSchierboom) [reply exercises/sgf-parsing/canonical-data.json]: Ignore this comment. For some reason I thought that there was another test case using escaped newlines. Sorry.

File: exercises/sgf-parsing/canonical-data.json (hunk 11 above)
Comment (ErikSchierboom) [reply exercises/sgf-parsing/canonical-data.json:343]: Absolutely.

**Review by ErikSchierboom — APPROVED** [review]

**Comment by ErikSchierboom** [comment]
CC @exercism/reviewers

**Comment by petertseng** [comment]
Please understand that it was necessary for me to rebase this because #1917 reformatted this exercise. My intent is that the only thing changed is whitespace. Believing this to be true (observe how `git diff -w 03d447f 427d136 exercises/sgf-parsing`), I believe this PR remains reviewable.

**Review by kytrinyx — APPROVED** [review]

**Review by glennj — APPROVED** [review]
